plugins/auth-backend/src/lib/OAuthProvider.ts

```diff
@@ -39,9 +40,41 @@ export type Options = {
   tokenIssuer: TokenIssuer;
 };
 
+const readState = (stateString: string): OAuthState => {
+  try {
+    const state = JSON.parse(decodeURIComponent(stateString));
+    return {
+      nonce: state.nonce,
+      env: state.env,
+    };
+  } catch (e) {
+    return {
+      nonce: undefined,
+      env: undefined,
+    };
+  }
+};
+
+export const encodeState = (state: OAuthState): string => {
+  return encodeURIComponent(JSON.stringify(state));
```

Comment:
Tbh I prefer URI encoding twice over JSON. The JSON delimiters end up really ugly and take a lot of space. It would actually be more compact to url-safe base64 encode this.

e.g.

```ts
const encoded = encodeURIComponent(new URLSearchParams(state).toString());
const decoded = Object.fromEntries(new URLSearchParams(decodeURIComponent(encoded)));
```

Reply:
For comparison it's

```
env%3Ddevelopment%26nonce%3D45c8g7yn452v4v
```
vs
```
%7B%22env%22%3A%22development%22%2C%22nonce%22%3A%2245c8g7yn452v4v%22%7D
```

Reply:
:+1:  for this suggestion. Will update it.